### PR TITLE
Chores

### PR DESCRIPTION
**Pull request by sorki**
- Fixes bunch of warnings produced by more recent GHC
- CI updates
- Bunch of random small fixes and improvements I'm discovering while debugging QuickCheck failures from #441 which are caused by the improved `Arbitrary Double` generator that now also generates exact zeroes.

### CODE REVIEW

File: Graphics/Implicit/Definitions.hs [hunk 1]
```diff
@@ -304,7 +308,7 @@ instance Show SymbolicObj2 where
     Circle r         -> showCon "circle"     @| r
     Polygon ps       -> showCon "polygon"    @| ps
     Rotate2 v obj    -> showCon "rotate"     @| v     @| obj
-    Transform2 m obj -> showCon "transform2" @| m     @| obj
```
Comment (julialongtin) [inline Graphics/Implicit/Definitions.hs:307]: so, you don't like "transform, transform3", but do like "rotate, rotate3"?

File: Graphics/Implicit/Export/Render/HandleSquares.hs [hunk 2]
```diff
@@ -3,6 +3,8 @@
 -- Copyright (C) 2016, Julia Longtin (julial@turinglace.com)
 -- Released under the GNU AGPLV3+, see LICENSE
 
+{-# LANGUAGE LambdaCase #-}
```
Comment (julialongtin) [inline Graphics/Implicit/Export/Render/HandleSquares.hs:6]: Why is this here?

File: Graphics/Implicit/ObjectUtil/GetImplicit2.hs [hunk 3]
```diff
@@ -65,12 +70,20 @@ getImplicit2 ctx (Rotate2 θ symbObj) =
         obj = getImplicit2 ctx symbObj
     in
         obj $ V2 (x*cos θ + y*sin θ) (y*cos θ - x*sin θ)
+-- ignore if zeroes, TODO(srk): produce warning
+-- TODO(srk): produce warning and ignore if we get a non-ivertible matrix
```
Comment (julialongtin) [inline Graphics/Implicit/ObjectUtil/GetImplicit2.hs:74]: invertible. :)

File: Graphics/Implicit/ObjectUtil/GetImplicitShared.hs [hunk 4]
```diff
@@ -56,7 +56,7 @@ getImplicitShared ctx (UnionR _ []) =
 getImplicitShared ctx (UnionR r symbObjs) = \p ->
   rminimum r $ fmap (flip (getImplicit' ctx) p) symbObjs
 getImplicitShared ctx (IntersectR _ []) =
-  getImplicitShared @obj ctx Full
+  getImplicitShared @obj ctx Empty
```
Comment (julialongtin) [inline Graphics/Implicit/ObjectUtil/GetImplicitShared.hs:59]: is this just wrong?

File: programs/extopenscad.hs [hunk 5]
```diff
@@ -19,7 +19,7 @@ import Graphics.Implicit (union, runOpenscad)
 import Graphics.Implicit.Definitions (ℝ)
 
 -- Use default values when a Maybe is Nothing.
-import Data.Maybe (fromMaybe, maybe)
```
Comment (julialongtin) [inline programs/extopenscad.hs:22]: don't like alphabetic order?

File: tests/Graphics/Implicit/Test/Instances.hs [hunk 6]
```diff
@@ -182,17 +234,65 @@ instance Observe () Double Double
...
 -- | Two 'SymbolicObj3's are the same if their 'getImplicit' functions agree at
 -- all points (up to an error term of 'epsilon')
-instance Observe (ℝ3, ()) Insidedness SymbolicObj3 where
-  observe p = insidedness . observe p . getImplicit
+instance Observe (ℝ3, ()) (TestResult SymbolicObj3 Double) SymbolicObj3 where
+  observe (sampledAt, _) obj = makeTestResult obj sampledAt
+
+------------------------------------------------------------------------------
+-- | The number of decimal points we need to agree to assume two 'Double's are
+-- equal.
```
Comment (julialongtin) [inline tests/Graphics/Implicit/Test/Instances.hs:273]: excuse me while i hurl. :)

**Review by julialongtin — APPROVED** [review]

File: Graphics/Implicit/Export/Render/HandleSquares.hs (hunk 2 above)
Comment (sorki) [reply Graphics/Implicit/Export/Render/HandleSquares.hs:6]: Required for `\case` bellow. I don't like to write `\x -> case x of`. Maybe we could enable some harmless extensions globally as well but per file is fine for now, although missing `ScopedTypeVariables` sometimes lead to a lot head scratching.

File: Graphics/Implicit/ObjectUtil/GetImplicitShared.hs (hunk 4 above)
Comment (sorki) [reply Graphics/Implicit/ObjectUtil/GetImplicitShared.hs:59]: I think so, based on https://en.wikipedia.org/wiki/Intersection_(set_theory)#Algebraic_properties 

> The intersection of any set with the [empty set](https://en.wikipedia.org/wiki/Empty_set) results in the empty set;

File: tests/Graphics/Implicit/Test/Instances.hs (hunk 6 above)
Comment (sorki) [reply tests/Graphics/Implicit/Test/Instances.hs:273]: I think that the reason it wasn't required was that `Arbitrary Double / Gen Double` wasn't generating small enough or close to zero things and transforms. Before, there was `epsilon = 5` and currently even `epsilon = 16` works quite well but eventually, with large enough `--qc-max-success` it breaks. This seems to hold up to `--qc-max-success=10M`

File: Graphics/Implicit/Definitions.hs (hunk 1 above)
Comment (sorki) [reply Graphics/Implicit/Definitions.hs:307]: Thanks, renamed to match `Rotate / Rotate3`.

We might also get rid of Rotate, Rotate3 and Translate and implement these using Transform/Transform3 as these are easier to compose (just a matrix multiplication), so we don't have redundant code paths to do the same thing. On a second thought, they wouldn't `Show` nicely so NO. Better to do that in canonicalization pass which won't affect `Show` of user defined trees.

**Comment by sorki** [comment]
I still want to factor out guards into canonicalization function, that will also solve it for `getBox2|3`.

File: Graphics/Implicit/Definitions.hs (hunk 1 above)
Comment (sorki) [reply Graphics/Implicit/Definitions.hs:307]: The naming was consistent, just the `showCon` param was outputting transform2 instead of transform (`transform = Transform2`). Restored.

File: Graphics/Implicit/ObjectUtil/GetImplicitShared.hs (hunk 4 above)
Comment (sorki) [reply Graphics/Implicit/ObjectUtil/GetImplicitShared.hs]: Also our implicit space (or starting object?) seems to be `emptySpace` - if we have `union [oneObject]` or just `oneObject` (without implicit union xD) it seems like it can only be unioned with `emptySpace`, because if it was a `fullSpace` we would get `fullSpace` instead of `oneObject`.

File: Graphics/Implicit/Export/SymbolicFormats.hs [hunk 7]
```diff
@@ -112,7 +112,7 @@ buildShared (WithRounding r obj) | r == 0 = build obj
 
 buildShared(UnionR _ _) = error "cannot provide roundness when exporting openscad; unsupported in target format."
 buildShared(IntersectR _ _) = error "cannot provide roundness when exporting openscad; unsupported in target format."
-buildShared(DifferenceR _ _ _) = error "cannot provide roundness when exporting openscad; unsupported in target format."
+buildShared(DifferenceR {}) = error "cannot provide roundness when exporting openscad; unsupported in target format."
```
Comment (julialongtin) [inline Graphics/Implicit/Export/SymbolicFormats.hs:115]: why quash the argument count just here?

File: Graphics/Implicit/ExtOpenScad/Eval/Expr.hs [hunk 8]
```diff
@@ -3,10 +3,6 @@
 -- Copyright (C) 2016, Julia Longtin (julial@turinglace.com)
 -- Released under the GNU AGPLV3+, see LICENSE
 
--- allow us to specify what package to import what module from.
--- We don't actually care, but when we compile our haskell examples, we do.
-{-# LANGUAGE PackageImports #-}
```
Comment (julialongtin) [inline Graphics/Implicit/ExtOpenScad/Eval/Expr.hs:8]: what made this redundant?

**Review by julialongtin — APPROVED** [review]

File: Graphics/Implicit/Export/SymbolicFormats.hs (hunk 7 above)
Comment (sorki) [reply Graphics/Implicit/Export/SymbolicFormats.hs:115]: Hlint complains about it when there are 3 and more

File: Graphics/Implicit/ExtOpenScad/Eval/Expr.hs (hunk 8 above)
Comment (sorki) [reply Graphics/Implicit/ExtOpenScad/Eval/Expr.hs:8]: No idea! (found by hlint).
